### PR TITLE
fix golint, go vet and go fmt

### PR DESCRIPTION
**Pull request by feliixx**
**WIP - Do not merge**
 
This PR is an attempt to fix `golint`, `go vet` and `gofmt` warning of all files in the project ( see #43 for details) 

A few warnings are still raised by `go vet`: 

```
session.go:661: assignment copies lock value to scopy: mgo.Session contains sync.RWMutex
```
   -> not sure how to handle this one 

```
internal/json/decode_test.go:1350: struct field m has json tag but is not exported
internal/json/decode_test.go:1351: struct field m2 has json tag but is not exported
internal/json/tagkey_test.go:56: struct field tag `:"BadFormat"` not compatible with reflect.StructTag.Get: bad syntax for struct tag key
```
  -> part of the tests and intended to be badly formatted. 

I was able to fix all golint warnings, but bad news is that it currently breaks backward compatibility...
`golint` raise warnings about `Id` not being `ID`, and theres many occurence of this in mgo (`ObjectId`, `FindId()` ...) 

This might be fixed with type alias to preserve backward compatibility ? 

Waiting for any feedback on this, 

Feliixx

### CODE REVIEW

**Comment by domodwyer** [comment]
Hey @feliixx 

I just had a quick look at the mutex thing - `copySession()` is creating a shallow copy of the struct (so same pointers, etc but fields referenced by value are copied). Essentially it's a lazy way of doing:

```
scopy := Session{
    defaultdb: session.defaultdb,
    sourcedb: session.sourcedb,
    ...
    m: sync.Mutex{},
    ...
}
```

So long as this is the only place a session is copied (and it looks like it is, otherwise you'd have more mutex copy errors) the above change should be fine.

I'm aware of the other two - and I really don't like what I'm about to suggest...

* For the struct tags, I think in this case [disabling](https://golang.org/cmd/vet/#hdr-Struct_tags) vet's struct tag checks might be best - we could do some fancy stuff with build tags but it's a bit hard to reason about. 

If the tags are malformed the code won't work, and this should be caught by any unit tests added alongside any new structs with tags.

* For the `Id => ID` conversion - again, sorry! - I'd rather keep backwards compatibility and not run `golint` as part of the CI - apparently it's not really supposed to be used for this purpose anyway (see [here](https://github.com/golang/lint#purpose)).

I don't think type aliasing will be able to help us with the ID thing but I'm definitely open to ideas - we'd have to do it when 1.9 is a little "older" - a lot of people will currently still be using 1.8 and earlier.

As always, genuinely grateful for the time you've put in contributing - **thanks!**

Dom

**Comment by feliixx** [comment]
Hey @domodwyer, 

Totally understand that keeping backward compatibility is more important. 

I removed all Id/ID related changes, and fixed the warning in `copySession()` 

So here is a solution : 

  - run `golint ./... | grep -v 'ID' | cat ` so warnings about Id/ID aren't printed, but all other are. If some warnings are found, the build does not fail

  - run `go vet ` on exported packages only (**mgo**, **mgo/bson** and **mgo/txn**). Content of `internal`  would not be tested . If warnings are found, the build fails 

Feliixx

**Comment by domodwyer** [comment]
Hey @feliixx 

Both are good solutions for me :) Any idea why tests on 3.0.X always fail? I've retried them a bunch of times.

Dom

**Comment by feliixx** [comment]
Hi @domodwyer, 

tests on 3.0.x were failing because the command `drop` doesn't return the same error as 2.6, 3.2 & 3.4 when trying to delete a collection that doesn't exist. 
It might be a bug, couldn' test this with 3.0.15 (last release for the 3.0.x serie). 
I updated the test `TestRepairCursor`, everything should be fine now!

**Comment by domodwyer** [comment]
Thanks @feliixx!

File: bson/bson.go [hunk 1]
```diff
@@ -83,13 +83,13 @@ const (
 	BinaryGeneric     byte = 0x00
 	BinaryFunction    byte = 0x01
 	BinaryBinaryOld   byte = 0x02
-	BinaryUUIDOld     byte = 0x03
-	BinaryUUID        byte = 0x04
+	BinaryUUIdOld     byte = 0x03
```
Comment (domodwyer) [inline bson/bson.go]: I think these got swept up in a find-and-replace - should still be `ID`

File: bson/bson.go [hunk 2]
```diff
@@ -189,17 +189,17 @@ type Raw struct {
...
+// RawDocElem elements of RawD type.
 type RawDocElem struct {
 	Name  string
 	Value Raw
 }
 
-// ObjectId is a unique ID identifying a BSON value. It must be exactly 12 bytes
+// ObjectId is a unique Id identifying a BSON value. It must be exactly 12 bytes
 // long. MongoDB objects by default have such a property set in their "_id"
 // property.
```
Comment (domodwyer) [inline bson/bson.go]: Accidental ID/Id replace (url broken)

File: bson/bson.go [hunk 3]
```diff
@@ -447,7 +447,7 @@ var Undefined undefined
 //   0x00 - Generic. This is decoded as []byte(data), not Binary{0x00, data}.
 //   0x01 - Function (!?)
 //   0x02 - Obsolete generic.
-//   0x03 - UUID
+//   0x03 - UUId
```
Comment (domodwyer) [inline bson/bson.go]: UUID

File: bson/bson_corpus_spec_test.go [hunk 4]
```diff
@@ -740,15 +740,15 @@ func (s *S) TestDBPointer_type__deprecated__String_not_null_terminated(c *C) {
 	testDecodeError(c, b, &resultD)
 }
 
-func (s *S) TestDBPointer_type__deprecated__short_OID__less_than_minimum_length_for_field_(c *C) {
+func (s *S) TestDBPointer_type__deprecated__short_OId__less_than_minimum_length_for_field_(c *C) {
```
Comment (domodwyer) [inline bson/bson_corpus_spec_test.go]: OID

File: bson/bson_corpus_spec_test.go [hunk 5]
```diff
@@ -740,15 +740,15 @@ func (s *S) TestDBPointer_type__deprecated__String_not_null_terminated(c *C) {
...
 	b, err := hex.DecodeString("160000000C61000300000061620056E1FC72E0C91700")
 	c.Assert(err, IsNil)
 
 	var resultD bson.D
 	testDecodeError(c, b, &resultD)
 }
 
-func (s *S) TestDBPointer_type__deprecated__short_OID__greater_than_minimum__but_truncated_(c *C) {
+func (s *S) TestDBPointer_type__deprecated__short_OId__greater_than_minimum__but_truncated_(c *C) {
```
Comment (domodwyer) [inline bson/bson_corpus_spec_test.go]: OID

File: bson/bson_corpus_spec_test.go [hunk 6]
```diff
@@ -1372,7 +1372,7 @@ func (s *S) TestObjectId_Random(c *C) {
 	testDecodeSkip(c, b)
 }
 
-func (s *S) TestObjectId_OID_truncated(c *C) {
+func (s *S) TestObjectId_OId_truncated(c *C) {
```
Comment (domodwyer) [inline bson/bson_corpus_spec_test.go]: OID

File: bson/bson_test.go [hunk 7]
```diff
@@ -580,25 +580,25 @@ func (s *S) TestUnmarshalNilInStruct(c *C) {
...
-	Other byte "name" // Tag should precede.
+	Other byte `bson:"name"` // Tag should precede.
 }
 
 var marshalErrorItems = []testItemType{
 	{bson.M{"": uint64(1 << 63)},
 		"BSON has no uint64 type, and value is too large to fit correctly in an int64"},
 	{bson.M{"": bson.ObjectId("tooshort")},
-		"ObjectIDs must be exactly 12 bytes long \\(got 8\\)"},
+		"ObjectIds must be exactly 12 bytes long \\(got 8\\)"},
```
Comment (domodwyer) [inline bson/bson_test.go]: ID

File: bson/encode.go [hunk 8]
```diff
@@ -275,7 +275,7 @@ func (e *encoder) addElem(name string, v reflect.Value, minSize bool) {
 		switch v.Type() {
 		case typeObjectId:
 			if len(s) != 12 {
-				panic("ObjectIDs must be exactly 12 bytes long (got " +
+				panic("ObjectIds must be exactly 12 bytes long (got " +
```
Comment (domodwyer) [inline bson/encode.go]: IDs

File: bson/encode.go [hunk 9]
```diff
@@ -417,7 +417,7 @@ func (e *encoder) addElem(name string, v reflect.Value, minSize bool) {
 			e.addElemName(0x0C, name)
 			e.addStr(s.Namespace)
 			if len(s.Id) != 12 {
-				panic("ObjectIDs must be exactly 12 bytes long (got " +
+				panic("ObjectIds must be exactly 12 bytes long (got " +
```
Comment (domodwyer) [inline bson/encode.go]: Ids

File: gridfs.go [hunk 10]
```diff
@@ -39,6 +39,26 @@ import (
 	"github.com/globalsign/mgo/bson"
 )
 
+// GridFS stores files in two collections:
+//
+// - chunks stores the binary chunks. For details, see The chunks Collection.
```
Comment (domodwyer) [inline gridfs.go]: `The` => `the`

File: gridfs.go [hunk 11]
```diff
@@ -376,11 +397,11 @@ func (file *GridFile) Id() interface{} {
 	return file.doc.Id
 }
 
-// SetId changes the current file Id.
+// setId changes the current file Id.
 //
 // It is a runtime error to call this function once the file has started
 // being written to, or when the file is not open for writing.
-func (file *GridFile) SetId(id interface{}) {
+func (file *GridFile) setId(id interface{}) {
```
Comment (domodwyer) [inline gridfs.go]: Is not exporting this intentional?

File: internal/json/example_test.go [hunk 12]
```diff
@@ -14,14 +14,15 @@ import (
 	"strings"
 )
 
+
 func ExampleMarshal() {
 	type ColorGroup struct {
-		ID     int
+		Id     int
```
Comment (domodwyer) [inline internal/json/example_test.go]: id

File: internal/json/example_test.go [hunk 13]
```diff
@@ -14,14 +14,15 @@ import (
...
 		Name   string
 		Colors []string
 	}
 	group := ColorGroup{
-		ID:     1,
+		Id:     1,
```
Comment (domodwyer) [inline internal/json/example_test.go]: id

File: internal/json/example_test.go [hunk 14]
```diff
@@ -31,7 +32,7 @@ func ExampleMarshal() {
 	}
 	os.Stdout.Write(b)
 	// Output:
-	// {"ID":1,"Name":"Reds","Colors":["Crimson","Red","Ruby","Maroon"]}
+	// {"Id":1,"Name":"Reds","Colors":["Crimson","Red","Ruby","Maroon"]}
```
Comment (domodwyer) [inline internal/json/example_test.go]: id

File: gridfs.go [hunk 15]
```diff
@@ -411,7 +432,7 @@ func (file *GridFile) ContentType() string {
 	return file.doc.ContentType
 }
 
-// ContentType changes the optional file content type.  An empty string may be
+// SetContentType changes the optional file content type.  An empty string may be
```
Comment (domodwyer) [inline gridfs.go]: This is a breaking change for users of GridFile - I agree this is what the name should be though.

Opinions @globalsign ?

File: internal/sasl/sasl.go [hunk 16]
```diff
@@ -26,7 +26,8 @@ import (
 	"unsafe"
 )
 
-type saslStepper interface {
+// Stepper interface for saslSession
+type Stepper interface {
```
Comment (domodwyer) [inline internal/sasl/sasl.go]: Intentionally exported?

File: log.go [hunk 17]
```diff
@@ -34,15 +34,15 @@ import (
 // ---------------------------------------------------------------------------
 // Logging integration.
 
-// Avoid importing the log type information unnecessarily.  There's a small cost
+// LogLogger avoid importing the log type information unnecessarily.  There's a small cost
 // associated with using an interface rather than the type.  Depending on how
 // often the logger is plugged in, it would be worth using the type instead.
-type log_Logger interface {
+type LogLogger interface {
```
Comment (domodwyer) [inline log.go]: Intentionally exported?

File: session.go [hunk 18]
```diff
@@ -44,23 +44,32 @@ import (
 	"github.com/globalsign/mgo/bson"
 )
 
+// Mode read preference mode. See Eventual, Monotonic and Strong for details
+//
+// Relevant documentation on read preference modes:
+//
+//     http://docs.mongodb.org/manual/reference/read-preference/
```
Comment (domodwyer) [inline session.go]: Thanks for adding all the reference URLs! Great addition 👍

File: session.go [hunk 19]
```diff
@@ -2416,7 +2499,7 @@ func (p *Pipe) Iter() *Iter {
 // When using MongoDB 3.2+ NewIter supports re-using an existing cursor on the
 // server. Ensure the connection has been established (i.e. by calling
 // session.Ping()) before calling NewIter.
-func (c *Collection) NewIter(session *Session, firstBatch []bson.Raw, cursorId int64, err error) *Iter {
+func (c *Collection) NewIter(session *Session, firstBatch []bson.Raw, CursorId int64, err error) *Iter {
```
Comment (domodwyer) [inline session.go]: accidental capitalisation (and in the documentation above)

File: session.go [hunk 20]
```diff
@@ -3828,11 +3917,11 @@ func (iter *Iter) Err() error {
 // a *QueryError type.
 func (iter *Iter) Close() error {
 	iter.m.Lock()
-	cursorId := iter.op.cursorId
-	iter.op.cursorId = 0
+	CursorId := iter.op.CursorId
```
Comment (domodwyer) [inline session.go]: Accidental capitalisation

File: session_test.go [hunk 21]
```diff
@@ -1355,20 +1356,20 @@ func (s *S) TestView(c *C) {
 	c.Assert(n, Equals, 2)
 
 	var result struct {
-		ID int    `bson:"_id"`
+		Id int    `bson:"_id"`
```
Comment (domodwyer) [inline session_test.go]: id

File: session_test.go [hunk 22]
```diff
@@ -1338,13 +1339,13 @@ func (s *S) TestView(c *C) {
 	c.Assert(names, DeepEquals, []string{"mycoll", "myview", "system.views"})
 
 	var viewInfo struct {
-		ID       string   `bson:"_id"`
+		Id       string   `bson:"_id"`
```
Comment (domodwyer) [inline session_test.go]: id

File: socket.go [hunk 23]
```diff
@@ -137,13 +137,13 @@ func (op *queryOp) finalQuery(socket *mongoSocket) interface{} {
 type getMoreOp struct {
 	collection string
 	limit      int32
-	cursorId   int64
+	CursorId   int64
```
Comment (domodwyer) [inline socket.go]: accidental capitalisation

File: socket.go [hunk 24]
```diff
@@ -137,13 +137,13 @@ func (op *queryOp) finalQuery(socket *mongoSocket) interface{} {
...
 	replyFunc  replyFunc
 }
 
 type replyOp struct {
 	flags     uint32
-	cursorId  int64
+	CursorId  int64
```
Comment (domodwyer) [inline socket.go]: accidental capitalisation

File: socket.go [hunk 25]
```diff
@@ -171,7 +171,7 @@ type deleteOp struct {
 }
 
 type killCursorsOp struct {
-	cursorIds []int64
+	CursorIds []int64
```
Comment (domodwyer) [inline socket.go]: accidental capitalisation

File: txn/txn.go [hunk 26]
```diff
@@ -137,7 +137,7 @@ func newNonce() string {
 
 type token string
 
-func (tt token) id() bson.ObjectId { return bson.ObjectIdHex(string(tt[:24])) }
+func (tt token) Id() bson.ObjectId { return bson.ObjectIdHex(string(tt[:24])) }
```
Comment (domodwyer) [inline txn/txn.go]: accidental capitalisation

File: txn/txn.go [hunk 27]
```diff
@@ -207,10 +207,13 @@ func (op *Op) name() string {
 }
 
 const (
-	// DocExists and DocMissing may be used on an operation's
+	// DocExists may be used on an operation's
 	// Assert value to assert that the document with the given
-	// Id exists or does not exist, respectively.
-	DocExists  = "d+"
+	// Id exists.
```
Comment (domodwyer) [inline txn/txn.go]: ID

File: txn/txn.go [hunk 28]
```diff
@@ -207,10 +207,13 @@ func (op *Op) name() string {
...
+	DocExists = "d+"
+	// DocMissing may be used on an operation's
+	// Assert value to assert that the document with the given
+	// Id does not exist.
```
Comment (domodwyer) [inline txn/txn.go]: ID

File: txn/txn.go [hunk 29]
```diff
@@ -291,12 +296,12 @@ var ErrAborted = fmt.Errorf("transaction aborted")
...
 // another transaction on any of the documents targeted by ops, as
 // long as the interruption was made after the transaction document
-// itself was inserted. Run Resume with the obtained transaction id
+// itself was inserted. Run Resume with the obtained transaction Id
 // to confirm whether the transaction was applied or not.
 //
 // Any number of transactions may be run concurrently, with one
 // runner or many.
-func (r *Runner) Run(ops []Op, id bson.ObjectId, info interface{}) (err error) {
+func (r *Runner) Run(ops []Op, Id bson.ObjectId, info interface{}) (err error) {
```
Comment (domodwyer) [inline txn/txn.go]: accidental capitalisation

File: txn/txn.go [hunk 30]
```diff
@@ -366,11 +371,11 @@ func (r *Runner) ResumeAll() (err error) {
 	return nil
 }
 
-// Resume resumes the transaction with id. It returns mgo.ErrNotFound
+// Resume resumes the transaction with Id. It returns mgo.ErrNotFound
 // if the transaction is not found. Otherwise, it has the same semantics
 // of the Run method after the transaction is inserted.
-func (r *Runner) Resume(id bson.ObjectId) (err error) {
-	t, err := r.load(id)
+func (r *Runner) Resume(Id bson.ObjectId) (err error) {
```
Comment (domodwyer) [inline txn/txn.go]: accidental capitalisation

File: txn/txn.go [hunk 31]
```diff
@@ -394,7 +399,7 @@ func (r *Runner) Resume(id bson.ObjectId) (err error) {
 //
 // Saved documents are in the format:
 //
-//     {"_id": <txn id>, <collection>: {"d": [<doc id>, ...], "r": [<doc revno>, ...]}}
+//     {"_id": <txn Id>, <collection>: {"d": [<doc Id>, ...], "r": [<doc revno>, ...]}}
```
Comment (domodwyer) [inline txn/txn.go]: accidental capitalisation

File: txn/txn.go [hunk 32]
```diff
@@ -481,11 +486,11 @@ func (r *Runner) PurgeMissing(collections ...string) error {
 	return nil
 }
 
-func (r *Runner) load(id bson.ObjectId) (*transaction, error) {
+func (r *Runner) load(Id bson.ObjectId) (*transaction, error) {
```
Comment (domodwyer) [inline txn/txn.go]: accidental capitalisation

File: txn/txn.go [hunk 33]
```diff
@@ -438,7 +443,7 @@ func (r *Runner) PurgeMissing(collections ...string) error {
 					found[txnId] = true
 					continue
 				}
-				logf("WARNING: purging from document %s/%v the missing transaction id %s", collection, tdoc.Id, txnId)
+				logf("WARNING: purging from document %s/%v the missing transaction Id %s", collection, tdoc.Id, txnId)
```
Comment (domodwyer) [inline txn/txn.go]: accidental capitalisation

File: txn/txn.go [hunk 34]
```diff
@@ -467,7 +472,7 @@ func (r *Runner) PurgeMissing(collections ...string) error {
 				found[txnId] = true
 				continue
 			}
-			logf("WARNING: purging from stash document %s/%v the missing transaction id %s", stdoc.Id.C, stdoc.Id.Id, txnId)
+			logf("WARNING: purging from stash document %s/%v the missing transaction Id %s", stdoc.Id.C, stdoc.Id.Id, txnId)
```
Comment (domodwyer) [inline txn/txn.go]: accidental capitalisation

File: txn/txn.go [hunk 35]
```diff
@@ -542,7 +546,7 @@ func valueNature(v interface{}) (value interface{}, nature typeNature) {
 	case reflect.Struct:
 		return v, natureStruct
 	}
-	panic("document id type unsupported by txn: " + rv.Kind().String())
+	panic("document Id type unsupported by txn: " + rv.Kind().String())
```
Comment (domodwyer) [inline txn/txn.go]: accidental capitalisation

File: gridfs.go (hunk 15 above)
Comment (feliixx) [reply gridfs.go]: The method name was already `SetContentType`, this is just a documentation change

File: internal/sasl/sasl.go (hunk 16 above)
Comment (feliixx) [reply internal/sasl/sasl.go]: yes, as it is returned by `New(username, password, mechanism, service, host string)` method which is exported ( [sasl.go](https://github.com/globalsign/mgo/blob/master/internal/sasl/sasl.go#L53) ) but maybe it shouldn't be ? I'm not familiar with this part of the driver, so can't tell...

File: bson/bson.go [hunk 36]
```diff
@@ -99,12 +99,12 @@ type Getter interface {
 	GetBSON() (interface{}, error)
 }
 
-// A value implementing the bson.Setter interface will receive the BSON
+// Setter interface : a value implementing the bson.Setter interface will receive the BSON
```
Comment (tadukurow) [inline bson/bson.go]: The space between "interface :" should be removed for consistency with the Getter comment. Very nitpicky this.

File: bson/bson.go [hunk 37]
```diff
@@ -130,10 +130,10 @@ type Setter interface {
 	SetBSON(raw Raw) error
 }
 
-// SetZero may be returned from a SetBSON method to have the value set to
+// ErrSetZero may be returned from a SetBSON method to have the value set to
 // its respective zero value. When used in pointer values, this will set the
 // field to nil rather than to the pre-allocated value.
-var SetZero = errors.New("set to zero")
+var ErrSetZero = errors.New("set to zero")
```
Comment (tadukurow) [inline bson/bson.go]: Potentially breaks backwards compat for little gain I'd say. @domodwyer, not sure how likely is this to be used?

File: gridfs.go (hunk 15 above)
Comment (tadukurow) [reply gridfs.go]: Doesn't look like an issue

File: gridfs_test.go [hunk 38]
```diff
@@ -141,12 +141,12 @@ func (s *S) TestGridFSFileDetails(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(info, IsNil)
 
-	file.SetId("myid")
+	// file.setId("myid")
```
Comment (tadukurow) [inline gridfs_test.go]: Any reason to keep this around?

File: internal/sasl/sasl.go (hunk 16 above)
Comment (tadukurow) [reply internal/sasl/sasl.go]: I agree the interface should be exported if returned from New.

File: gridfs.go (hunk 15 above)
Comment (domodwyer) [reply gridfs.go]: My bad, thought it was the method that changed

File: bson/bson.go (hunk 37 above)
Comment (domodwyer) [reply bson/bson.go]: I don't think this is much of an issue, it's unlikely to be used by 99% of people.

Comment (medicalwei) [reply bson/bson.go]: FYI, I am one of the 1% that's affected by this change. Add reactions here for "me too".

**Pull request merged**